### ISSUE TITLE
Prevent navigating twice when clicking a button quickly

### ISSUE THREAD

**Issue by nihgwu** [react-community/react-navigation#271]
When we double-click(or multi-click) a link quickly, it will navigate to the same screen twice, so when we want to goBack, we need click the back button twice to go back to parent screen, that's insane.

I use `ExperimentalNavigation` before, and if I double-click a link quickly, the app crashes because we are pushing a route with the same key, which is not allowed by `ExperimentalNavigation`, you can try the `UIExplorer` to reproduce, so I make a check for it:
1. newKey == latestKey, ignore this action
2. routes.contains(newKey), jumpTo(newKey)

Now in `react-navigation` it won't crash because every time we dispatch a new action, we get a different key

So my propose: 
1. don't generate key if provided in action when dispatching and check for existing
2. generate key according to route params, but I'm afraid that will be buggy because the screen relies on more params expect routeParams

Any thoughts on this?

**Comment by nihgwu** [comment]
```
export function routerReducer(state, action) {
  if (action.type.startsWith('Navigation/')) {
    const { type, routeName } = action
    const lastRoute = state.routes[state.routes.length - 1]
    if (type == lastRoute.type && routeName == lastRoute.routeName) return state
  }
  return AppNavigator.router.getStateForAction(action, state)
}
```
workaround for my case, I know we can custom nearly everything, but IMO this issue should be fixed in core

**Comment by satya164** [comment]
Another option is to just debounce the function you're dispatching from.

cc @ericvicenti

**Comment by nihgwu** [comment]
but debounce would make transition a bit laggy and we should let user to dispatch them own keys
`dispatch(NavigationActions.navigate({routeName: 'Detail', key: 'user-123'}))`

**Comment by grabbou** [comment]
> Now in react-navigation it won't crash because every time we dispatch a new action, we get a different key

To me it's a feature. Imagine a screen, like `userProfile` that is generic and accepts user details to display as a JSON. It can be in the stack as many times as I want.

**Comment by nihgwu** [comment]
@grabbou But I don't think we should override the provided key, at least show a warn about overriding or key generation strategy

**Comment by satya164** [comment]
> @nihgwu : but debounce would make transition a bit laggy

Maybe I'm missing something, but why would it?

cc @ericvicenti What do you think about this? This is has been a common issue I have faced in all of the apps I have worked on so far.

Mostly I used `NavigationExperimental` or `ExNavigation`. In case of former I did [changes in reducer to not push same routes with same params twice](https://github.com/satya164/PocketGear/blob/0f0b0a483928fd9f55ef7772c02c9a64c53adc69/src/components/Navigation/NavigationRoot.js#L92)

ExNavigation had a debouncing mechanism to prevent navigating twice in quick succession. cc @skevy

**Comment by nihgwu** [comment]
@satya164 my mistake, a debounce would be fine, but we should stop overriding custom keys

**Comment by arnsa** [comment]
@satya164 @nihgwu care to explain how can I implement this with a TabNavigator? Because state.routes[state.routes.length - 1] is always going to be last tab bar item for me.

**Comment by nihgwu** [comment]
I take the control of the default action flow of react-navigation, and apply a debounce, here is an example using dva:
https://github.com/nihgwu/react-native-dva-starter/blob/master/app/models/router.js#L24-L33

**Comment by joonhocho** [comment]
@nihgwu @satya164 @grabbou 
I am facing the same issus, and I believe everyone who uses this library will eventually have to address this issue.

I created https://github.com/react-community/react-navigation/issues/768, but closed as duplicate as I found this thread.
Here's my original post and a possible idea to fix this.


----

Pressing on the navigation header right button quickly for multiple times causes navigating to the new route multiple times.

I think it should be easy to prevent this behavior as it can be very common behavior that we do not want.

Possible solutions would be

Add transitioning property to navigation prop so that header button can disable itself if there is another navigation transition taking in place.
navigationOptions: {
  headers: ({navigate, transitioning}) => ({
      right:
        <Button
          icon="ios-settings-outline"
          disabled={transitioning}
          onPress={() => navigate('Settings')}
        />,
    })
},

**Comment by satya164** [comment]
Yes, it's in the v1 roadmap #723

**Comment by joonhocho** [comment]
@satya164 What do you think about adding a new property, `transitioning`?
It can be useful when you want to disable multiple buttons as transition occurs.

**Comment by satya164** [comment]
A prop will cause a re-render which will make transition slower. You can use `InteractionManager.runAfterInteractions` instead

**Comment by evisong** [comment]
I guess a `lodash.debounce()` can solve this already?

**Comment by dzuncoi** [comment]
Currently I create a function to override `navigate` from `addNavigationHelpers` as below:
```javascript
_addNavigationHelpers(navigation) {
    const original = addNavigationHelpers(navigation);
    let debounce;
    return {
      ...original,
      navigateWithDebounce: (routeName, params, action) => {
        let func = () => {
          clearTimeout(debounce);
          debounce = setTimeout(() => {
            navigation.dispatch(NavigationActions.navigate({
              routeName,
              params,
              action
            }));
          }, 200)
        }
        return func();
      }
    }
  }
```

And
```javascript
<VocabTabs navigation={this._addNavigationHelpers({
            dispatch: this.props.dispatch,
            state: this.props.nav,
          })}/>
```

Then you can use `this.props.navigation.navigateWithDebounce` instead of `this.props.navigation.navigate`

This, of course is a temporary approach. It'd better to have a life-cycle hook to detect when screen is in transition progress then prevent pushing another route.

**Comment by microwavesafe** [comment]
I used your method to good effect. I changed it around slightly because I needed long delays (page transitions can be quite slow) and I didn't want to delay the navigation.

```JS
    _addNavigationHelpers = (navigation) => {
        const original = addNavigationHelpers(navigation);
        let debounce;
        return {
            ...original,
            navigateWithDebounce: (routeName, params, action) => {
                let func = () => {
                    if (debounce) {
                        return;
                    }

                    navigation.dispatch(NavigationActions.navigate({
                        routeName,
                        params,
                        action
                    }));

                    debounce = setTimeout(() => {
                        debounce = 0;
                    }, 1000)
                };
                return func();
            }
        }
    };
```

**Comment by dzuncoi** [comment]
Nice approach @microwavesafe 👍

**Comment by cosivox** [comment]
@microwavesafe Where do I need to put that to get it working?

**Comment by dzuncoi** [comment]
@cosivox wherever you use `addNavigationHelpers`, replace it with the this one.

**Comment by cosivox** [comment]
@dzuncoi Do you mean inside library files? Because I don't use addNavigationHelpers anywhere in my code

**Comment by dzuncoi** [comment]
@cosivox if you don't use `addNavigationHelpers` then you can override `NavigationActions.navigate` function, no need to edit lib files

**Comment by ericvicenti** [comment]
Lets go with this approach, which solves the same problem:  https://github.com/react-community/react-navigation/issues/135

**Comment by Doko-Demo-Doa** [comment]
@dzuncoi May I ask for an example with `NavigationActions.navigate` ? I can't seem to find it in this repo.

**Comment by dzuncoi** [comment]
@Doko-Demo-Doa 
You can find it [here](https://github.com/react-community/react-navigation/blob/master/src/NavigationActions.js)

It's a basic action, you can use it as below:
```javascript
NavigationActions.navigate({
    routeName,
    params,
    action
})
```

**Comment by Doko-Demo-Doa** [comment]
Thank you, but what about `this.props.navigation.navigate` ? Right now I don't use addNavigationHelpers (because I don't use redux, for now) so does that mean I have to modify the lib?

**Comment by dzuncoi** [comment]
As I said, you can override `navigate` action, no need to modify libs.
Something like:
```javascript
NavigationActions.overridedNavigate = (routeName, params, action) => {
    // some override logic
    return NavigationActions.navigate(routeName, params, action)
}
```

**Comment by Doko-Demo-Doa** [comment]
I finally found a way to solve my problem. I had to edit the lib after all because I use this.props.navigation.navigate a lot and I don't want to modify every single line of that.

Edit the `node_modules/react-navigation/src/addNavigationHelpers.js`
````
/**
 * @flow
 *
 * Helpers for navigation.
 */

import type {
  NavigationAction,
  NavigationProp,
  NavigationParams,
} from './TypeDefinition';

import NavigationActions from './NavigationActions';

export default function<S: *> (navigation: NavigationProp<S, NavigationAction>) {
  let debounce = true; // Add this.
  return {
    ...navigation,
    goBack: (key?: ?string): boolean => navigation.dispatch(NavigationActions.back({
      key: key === undefined ? navigation.state.key : key,
    })),
    navigate: (
      routeName: string,
      params?: NavigationParams,
      action?: NavigationAction): boolean => {
        // And this conditional check.
        if (debounce) {
          debounce = false;
          navigation.dispatch(NavigationActions.navigate({
            routeName,
            params,
            action,
          }));
          setTimeout(() => {
            debounce = true;
          }, 600);
        }
      },
      // End check
    /**
     * For updating current route params. For example the nav bar title and
     * buttons are based on the route params.
     * This means `setParams` can be used to update nav bar for example.
     */
    setParams: (params: NavigationParams): boolean =>
      navigation.dispatch(NavigationActions.setParams({
        params,
        key: navigation.state.key,
      })),
  };
}
````

**Comment by ugiacoman** [comment]
Was a solution ever officially implemented or do we still need to use a workaround?

**Comment by BradRyan** [comment]
I would prefer the default would be to prevent navigating to the same view twice if all the route parameters are unchanged.  I can't think of a scenario where rendering the same view (w/ same parameters) twice, as is the case w/ double tap, would ever be desirable.

**Comment by m1ckm** [comment]
I will give you time to rectify but get rid of the shit shoot and gay reference stuff out of there.

**Comment by JodiWarren** [comment]
Has anyone successfully implemented this when using Redux action creators?

I have an app using a push action of this style, mostly for NavigationExperimental compatibility:

```javascript
export const push: Action = (route) => {
    return {
        type: 'Navigation/NAVIGATE',
        routeName: route.template,
        params: {
            ...route,
        },
    };
};
```

Then my connected components receive this action in from `mapDispatchToProps`.

I can't think of a good place to implement a debounce without manually passing the navigation prop all over the place and rewriting all of my `push` actions.

**Comment by hydraZty** [comment]
@dzuncoi Hi bro ,Where do I add this code?

**Comment by dzuncoi** [comment]
@hydraZty Do you use Redux. If yes, then whenever you use `addNavigationHelpers` as this [docs](https://reactnavigation.org/docs/guides/redux) says, just replace it by your customize function, better using *microwavesafe* method than mine.

**Comment by hydraZty** [comment]
No, i didn't use REDUX. I use react-native just few days，and i don't understand the react-navigation how to route my screen, like @Doko-Demo-Doa , i use `this.props.navigation.navigate`,what should i do

**Comment by hydraZty** [comment]
And i don't want to change my libs code，thanks for your help @dzuncoi

**Comment by dzuncoi** [comment]
If so, you can create an override function, like my [comment here](https://github.com/react-community/react-navigation/issues/271#issuecomment-295581754)
Then instead of using `this.props.navigation.navigate`, you should use `this.props.navigation. overridedNavigate`. I didn't test this approach, but hope it will work :)

**Comment by hydraZty** [comment]
```
        let debounce;
        if (debounce) return;
        this.props.navigation.navigate(location);
        debounce=setTimeout(() =>{
            debounce=0;
        }, 800);
```
This is my solution.....easy but repeat everywhere

**Comment by wellyshen** [comment]
@dzuncoi Nice man, I simplify your solution as below, it works cool :

```javascript
import { addNavigationHelpers, NavigationActions } from 'react-navigation';
import _ from 'lodash';

export const _addNavigationHelpers = (navigation: Object) => {
  const original = addNavigationHelpers(navigation);

  return {
    ...original,
    navigateWithDebounce: _.debounce((routeName, params, action) => {
      navigation.dispatch(
        NavigationActions.navigate({
          routeName,
          params,
          action,
        }),
      );
    }, 500),
  };
};
```

**Comment by nihgwu** [comment]
@wellyshen what if I want to dispatch the Navigation action directly? I think the right way is to handle the debounce in reducers, see https://github.com/react-community/react-navigation/issues/271#issuecomment-286973106

**Comment by Palisand** [comment]
I am using **nested** navigators and based my workaround on BradRyan's and nihgwu's comments.

``` javascript
const Index = StackNavigator({...});

const App = StackNavigator({
  Index: {
    screen: Index
  },
  ...
});

const navigateOnce = (getStateForAction) => (action, state) => {
  const {type, routeName} = action;
  return (
    state &&
    type === NavigationActions.NAVIGATE &&
    routeName === state.routes[state.routes.length - 1].routeName
  ) ? null : getStateForAction(action, state);
};

App.router.getStateForAction = navigateOnce(App.router.getStateForAction);
Index.router.getStateForAction = navigateOnce(Index.router.getStateForAction);
```

**Comment by wellyshen** [comment]
@nihgwu Yea, you are right, forget it.

**Comment by giautm** [comment]
@Palisand 👍

**Comment by nightspirit** [comment]
Here is my hack to solve this issue. In my setup, I was hooking the navigator to redux. I simply  throttle the dispatch function with lodash.throttle and it works for my need.

``` javascript
class App extends React.Component {
  render() {
    return (
      <AppNavigator navigation={addNavigationHelpers({
        dispatch: _.throttle(this.props.dispatch, 300),
        state: this.props.nav,
      })} />
    );
  }
}
```

The idea is give it some throttling in between screen transition so navigate/back action won't fire multiple times from UI. Since they all rely on dispatch function so I figure it probably the best place to apply the throttling.

**Comment by gregblass** [comment]
Is there a recommended solution to this currently?

**Comment by Asleepace** [comment]
A simple solution that worked for me was adding a flag on the component that triggers the navigation, and toggling it when navigating. Then using setTimeout like this:

      setTimeout(this.toggleNavigation.bind(this),500);

to toggle the flag back to normal once it triggers navigation.

**Comment by wellyshen** [comment]
@Asleepace,

Here's my solution, I'm using redux :

```javascript
/* @flow */

import { NavigationActions } from 'react-navigation';
import _ from 'lodash';

import type { Nav, Action } from '../types';
import Navigator from '../navigator/configNavigator';

type State = Nav;

// Used for getting current route from container
function getCurrentRoute(state) {
  if (state.routes) return getCurrentRoute(state.routes[state.index]);

  return state;
}

// Prevent screen called multitple times by quickly tap
const navigateOnce = getStateForAction => (action, state) => {
  const { type, routeName } = action;
  return state &&
    type === NavigationActions.NAVIGATE &&
    routeName === state.routes[state.routes.length - 1].routeName
    ? state
    : getStateForAction(action, state);
};

Navigator.router.getStateForAction = navigateOnce(Navigator.router.getStateForAction);

export default (state: State, action: Action): State => {
  const nextState = Navigator.router.getStateForAction(action, state);

  return _.assign({}, state, { ...nextState, getCurrentRoute: getCurrentRoute(nextState) });
};
```

**Comment by gregblass** [comment]
@Asleepace Would you mind giving a more in depth code explanation/example when you get a chance? I'm still not connecting the dots as to how you'd implement that.

**Comment by Asleepace** [comment]
@gregblass sure here is the more in depth solution:

`  constructor(props) {
    super(props);
    this.state = {
      isNavigating: false,
    }
  }
  viewPadlet() {
    if (this.state.isNavigating == false) {
      this.state.isNavigating = true;
      this.props.navigate('View',{ padlet:this.props.padlet });
      setTimeout(this.toggleNavigation.bind(this), 500);
    }
  }
  toggleNavigation() {
    this.state.isNavigating = false;
  }`

**Comment by slyfox4** [comment]
Is there is any solution to prevent second navigation when we click on "Back" button?

**Comment by giautm** [comment]
@Asleepace : I don't like that solution (i have tried before). I have 4 component can trig a navigation on a screen. So, i dont want to duplicate over and over. @Palisand 's solution is better right now (hopefully it will be official via an option like `navigateOne: true`)

**Comment by ugiacoman** [comment]
@Palisand You are a god among men 🥇

**Comment by mikelambert** [comment]
@Palisand After digging through the problems with nested navigators and trying to get it working with `debounce`, I stumbled on your solution. But I needed to replace `null` with `state` inside your `navigateOnce` call to make it work for me. Curious why you chose to return `null`...I'm using `react-navigation@1.0.0-beta.11` (in case their API/semantics changed).

**Comment by Palisand** [comment]
@mikelambert I chose to return `null` because I based my solution on [this](https://reactnavigation.org/docs/routers/#Blocking-Navigation-Actions) (the link was originally provided by @BradRyan [here](https://github.com/react-community/react-navigation/issues/271#issuecomment-299364240)). True, that particular example pertains to `NavigationActions.BACK` rather than `NAVIGATE`, but it worked for me so I thought I might as well post it here. @wellyshen's solution uses my `navigateOnce` but with the same change you have implemented. I will update my answer with the suggestion to replace `null` with `state`. What exactly wasn't working when you used `null`, were you still navigating twice?

**Comment by mikelambert** [comment]
I was doing a double-navigate to the same routeName, and my redux state was obliterated and replaced with `null`, then triggering errors elsewhere down the line that operated off the state.

Though I was doing a TabNavigator containing StackNavigators (with `navigateOnce` applied), not a StackNavigator of StackNavigators like in your example, which I think makes the difference. The parent `StackRouter` does this code:
```
          const route = childRouter.getStateForAction(action, childRoute);
          if (route === null) {
            return state; // What your code triggers
          }
          if (route && route !== childRoute) {
            return StateUtils.replaceAt(state, childRoute.key, route); // What my fix would trigger, I think this should work still...?
          }
```

Whereas the `TabRouter` has this code:
```
        const activeTabState = activeTabRouter.getStateForAction(
          action.action || action,
          activeTabLastState
        );
        if (!activeTabState && inputState) {
          return null; // What your code triggers, blowing up my redux state
        }
          if (route && route !== childRoute) {
            return StateUtils.replaceAt(state, childRoute.key, route); // What my fix would trigger, which again, works fine for me...
          }
```

(I filed a bug about another instance of `return null` here: https://github.com/react-community/react-navigation/issues/1852 , but I think I should file a new bug to mention the case above, since it feels like a bad divergence between StackRouter and TabRouter)

**Comment by levsero** [comment]
@Palisand, @mikelambert I was using nested stack navigators and if you replace `null` with `state` this does not prevent double navigation when routing inside the nested navigator.

I also had to update my redux reducer to properly handle `null` state returns, which is inline with the docs (which seem to expect the value to be null no occasion) https://reactnavigation.org/docs/guides/redux#Redux-Integration:

```
const nextState = ApplicationNavigator.router.getStateForAction(
    action,
    state
  );

 return nextState || state;
```

**Comment by CiralaDevelopment** [comment]
Another approach to the issue, only allows navigating once per 200 ms.

```
let timeout = null;

export const navigationReducer = (state, action) => {
    if (action.type.startsWith('Navigation/')) {
        if (timeout) {
            return state;
        }

        timeout = setTimeout(() => timeout = null, 200);

        const { routeName } = action;
        const lastRoute = state.routes[state.routes.length - 1];
    }

    const newState = router.getStateForAction(action, state);
    return (newState ? newState : state);
};

**Comment by vandaics** [comment]
I saw above suggest but none of them work for me. I have a list of items at the Left drawer, and user can click on each item to open the screen `Profile`. I also have a `screenA` which has a `buttonX` to show `Profile`. I want to prevent prevent display multiple same-`Profile` when user rapidly taps on `buttonX`. So what I do is:

**Note**: I use **redux**.

The workaround is:
- Any forward navigation, I will add a time in ms at params, it's named `navigateAt`
- Override getStateForAction get the previous route info and do a simple check.

The reducer:
```
const initialState = AppNavigator.router.getStateForAction(AppNavigator.router.getActionForPathAndParams('Splash'))
export default NavReducer = (state = initialState, action) => {
    let nextState
    switch (action.type) {
        case "ACTIONS.GO_BACK":
           // we don't prevent back so we don't need any wire thing here
            nextState = AppNavigator.router.getStateForAction(NavigationActions.back(), state)
            break
        case "OTHER_CASE": break
        case "ACTIONS.GOTO_PROFILE":
            
            nextState = AppNavigator.router.getStateForAction(NavigationActions.navigate({
                routeName: 'Profile',
                params: {
                      ...action.params,
                      navigateAt: new Date().getTime() // HERE IS THE KEY. ANY ACTION SHOULD HAVE IT
                }
            }), state)
            break
        default:
            nextState = AppNavigator.router.getStateForAction(action, state)
            break
    }
    return nextState || state
}
```
And the router
```
const App = StackNavigator({...})
const preventDoubleClick = (getStateForAction) => (action, state) => {
  const {type, routeName, params} = action;
  const preRoute = state.routes[state.routes.length - 1]
  return (
    state &&
    type === NavigationActions.NAVIGATE &&
    routeName === preRoute.routeName &&
    params.navigateAt - preRoute.params.navigateAt < 500 // PREVENT IN 500ms
  ) ? null : getStateForAction(action, state);
};

App.router.getStateForAction = preventDoubleClick(App.router.getStateForAction);
```

**Comment by Clowning** [comment]
@Palisand Your solution work great but in case, with a tabbar, i can't tap on my last tab :/ !

**Comment by giautm** [comment]
@Clowning : I think that solution only use for StackNavigator? :/

**Comment by Clowning** [comment]
@giautm  Yes i'm stupid sorry ahah

**Comment by Charles-LMQ** [comment]
guys, have you tried goBack and reset, there still "yellow bar" comes out, especially if you have a Modal on the next View.

**Comment by gregblass** [comment]
A little birdie told me that this may be prioritized soon with the V1 push.

**Comment by carllippert** [comment]
I have been having decent luck with this if your not hooking up to redux.

Inspired from stack overflow: https://stackoverflow.com/a/41215941/2784950

```javascript
import _ from 'lodash';

class HeaderButton extends Component {
  constructor(props) {
    super(props);
   this.debouncedOnpress = _.debounce(this._onPressFunc, 2000, {'leading': true, 'trailing': false,});
  }

  _onPressFunc = () => {
        this.props.onPress();
        console.log("Debounced!");
  } 

  render() {
    return (
        <TouchableOpacity
            onPress={this.debouncedOnpress}
            style={styles.button}>
            <Image source={require('../images/plus.png')}
            style={styles.image}/>
        </TouchableOpacity>
  );
 }
}
```
Excited for the fix in core!

**Comment by agarparaNit** [comment]
Here is my hack using setTimeout

`<TouchableOpacity disabled={this.state.disable} onPress={() => {this.navFunction();navigate('MobileNumber');}}>`

And here first disabling the touchableopacity then enabling it using setTimeout

```
navFunction() {
    this.setState({disable : true});
    setTimeout(() => {this.setState({disable: false})}, 2000)
  }
```

**Comment by nicolls1** [comment]
+1 please fix.

**Comment by gregblass** [comment]
Here's the workaround I came up with until this is addressed. I'm storing a `navigating` boolean in my redux store:

action:
```
export const setNavigating = (value) => ({
  type: SET_NAVIGATING,
  payload: value
})
```

reducer:
```
import {
  SET_NAVIGATING,
} from '../actions/types'

const INITIAL_STATE = false

export default (state = INITIAL_STATE, action) => {
  switch (action.type) {
    case SET_NAVIGATING:
      return action.payload
    default:
      return state
  }
}
```

nav util:
```
// Workaround for React Navigation's double tap bug

import { store } from '../store'
import { setNavigating } from '../actions'

debounceNav = (navigation, screen, options) => {
  let isNavigating = store.getState().navigating

  if (isNavigating == false) {
    store.dispatch(setNavigating(true))

    requestAnimationFrame(
      () => navigation.navigate(screen, options)
    )

    setTimeout(
      () => store.dispatch(setNavigating(false)),
      500
    )
  }
}

export { debounceNav }
```

Then I refactored every place in my application where `navigation.navigate` was happening with `debounceNav(navigation, 'screen-name', options)`.

Works for now. Looking forward to the official fix and being able to remove all that.

**Comment by listatt** [comment]
@microwavesafe Thanks!

**Comment by Clowning** [comment]
@Palisand I've used your system in my none-redux app and it's works like a charm but after added Redux i've an error : 

'Cannot read property 'router' of unedfined' 

Do you have any idea?

I've changed null to state but still happening!

**Comment by sun2rise** [comment]
This is the solution i came up with: 
- dig recursively into the `NavigationState` until I get the current `routeName`,
- then I do the same for the **next** `NavigationState` and get the **next** `routeName`.
- If it's the same route I simply return the current state, otherwise i let the navigation happen

This is the smallest modification needed and works with nested navigators too!

navigation_reducer.js
```js
import { AppNavigator } from './../AppNavigator';

const getCurrentRouteName = (state) => {
  const route = state.routes[state.index];
  return typeof route.index === 'undefined' ? route.routeName : getCurrentRouteName(route); 
}

export default (state, action) => {
  const nextState = AppNavigator.router.getStateForAction(action, state);

  // prevents navigating twice to the same route
 if (state && nextState) {
    const stateRouteName = getCurrentRouteName(state);
    const nextStateRouteName = getCurrentRouteName(nextState);
    return stateRouteName === nextStateRouteName ? state : nextState;
  }

  // Simply return the original `state` if `nextState` is null or undefined.
  return nextState || state;
};
``` 

What do you think?

**Comment by samw2k00** [comment]
There's some overkill solutions and (not so nice) workarounds, for what should be a simple fix to the core lib.

**My problem with the previously provided solutions:** the current state hasn't changed yet at the point of interception - meaning I can't compare newState vs prevState.

Ideally the library would provide an interceptor, which would allow the developer to provide custom logic, which allows/rejects the [`_onNavigationStateChange`](https://github.com/react-community/react-navigation/blob/master/src/createNavigationContainer.js#L180).

Until that solution, I've hacked in the following code (to prevent adding two routes of the same name onto the stack, in a row).

[Before this line](https://github.com/react-community/react-navigation/blob/master/src/createNavigationContainer.js#L179)... added this code:

```
const prevNav = state.nav
if (prevNav.routes[prevNav.index].routeName == action.routeName) {
  console.log('block double navigation to the same route - routeName=' + action.routeName)
  return false
}
```

**Comment by sun2rise** [comment]
@samw2k00 are you referring to my solution (the comment before yours)?
Are you using that code as reducer of Redux?

**Comment by samw2k00** [comment]
@sun2rise  sorry, i was referring to the solution proposed by @satya164 @nihgwu 

where you implement using redux and monitor the previous state. 

i had the same problem as @arnsa where the route state keep showing the last tab bar item.

> care to explain how can I implement this with a TabNavigator? Because state.routes[state.routes.length - 1] is always going to be last tab bar item for me.
> --

**Comment by adam-denny** [comment]
Expanding on @sun2rise's answer, this works for me using the navigator as a view component:

```
<MainTabNavigator 
  onNavigationStateChange={(prevState, currentState, action) => {
    
    const defaultGetStateForAction = MainTabNavigator.router.getStateForAction;
    
    const getCurrentRouteName = (state) => {
      const route = state.routes[state.index];
      return typeof route.index === 'undefined' ? route.routeName : getCurrentRouteName(route);
    }
    
    MainTabNavigator.router.getStateForAction = (action, state) => {
      if(action.routeName == getCurrentRouteName(state)){
        return null;
      }
      return defaultGetStateForAction(action, state);
    };
  
  }}
/>
```

**Comment by oregano1201** [comment]
Hi all, I am new to React-Native so many things I don't quite understand.
I create a custom workaround on the button behavior to prevent double click.

Here is my code, hopefully you can find it useful.

function _safeClick(obj,ref,evt){
  //maintain an array of last click time stamp
  if(!obj.lastClick) { obj.lastClick = Array();}

  //check if array with ref as key exist (ever been click before)
  if(!obj.lastClick[ref]) { obj.lastClick[ref] = new Date(0); }

  //trigger event when at least 1000ms time difference since last click
  if((new Date()).getTime() - obj.lastClick[ref].getTime() > 1000){
    evt();
    obj.lastClick[ref]=new Date();
  }
}

and here is how I use the function in a TouchableOpacity component

          <TouchableOpacity onPress={()=>{_safeClick(this,'btn1', ()=>{navigate('Page1');})}} >
            <Text >Go To Page 1</Text>
          </TouchableOpacity>

          <TouchableOpacity onPress={()=>{_safeClick(this,'btn2', ()=>{navigate('Page2');})}} >
            <Text >Go To Page 2</Text>
          </TouchableOpacity>

**Comment by jamsch** [comment]
I had problems implementing Palisand's comment because I had nested navigators. Here's my solution for Redux users.

**reducers/nav.js**
```javascript
import { NavigationActions } from 'react-navigation';
import AppNavigator from '../../screens';

// Gets the current route name
function getCurrentRouteName(state) {
  if (!state || !state.index) return state.routeName;
  return getCurrentRouteName(state.routes[state.index]);
}

const navReducer = (state = initialState, action) => {
  const { type, routeName } = action;
  const routeHasChanged = Boolean(action.routeName !== getCurrentRouteName(state))
  // Return current state if no routes have changed
  if (type === NavigationActions.NAVIGATE && routeHasChanged) {
    return state;
  }
  // Else return new navigation state or the current state
  return AppNavigator.router.getStateForAction(action, state) || state;
}

export default navReducer;
```

**Comment by tomaskazatel** [comment]
Thank you @jamsch redux fix works perfect but I think you have a mistake in the condition on this line `if (!routeHasChanged) {`

**Comment by jamsch** [comment]
@tomaskazatel Whoops my bad. I used a '===' instead of a '!==' on the previous line to indicate that the route name was different.

**Comment by rgabs** [comment]
I fixed this bug by creating a module which calls a function only once in the passed interval.

Example: If you wish to navigate from Home -> About
And you press the About button twice in say `400 ms`.

```js
navigateToAbout = () => dispatch(NavigationActions.navigate({routeName: 'About'}))

const pressHandler = callOnce(navigateToAbout,400);
<TouchableOpacity onPress={pressHandler}>
 ...
</TouchableOpacity>
```
The module will take care that it calls navigateToAbout `only once` in 400 ms.

Here is the link to the NPM module: https://www.npmjs.com/package/call-once-in-interval

**Comment by jamsch** [comment]
FYI I found an issue with my solution when navigating between different nested navigators. I've updated the code with a new function `getActionRouteName` to solve this issue.

**Comment by parakhod** [comment]
I just wrote the redux middleware for react-navigation debouncing, https://github.com/parakhod/react-navigation-redux-debouncer

Just add it and it will block all the unnecessary navigation actions

**Comment by wellyshen** [comment]
Hi Guys, update my [code](https://github.com/react-community/react-navigation/issues/271#issuecomment-307607245) for handle this issue with redux. Try it and tell me if it works for your scenario or not ;)

**Comment by dktan** [comment]
@Palisand your so helpful

**Comment by blakek** [comment]
Just curious—is this marked as closed because #135 is open?  Thanks!

**Comment by pencilcheck** [comment]
Thanks to @Palisand's workaround, I got inspired to improve upon and made another workaround that works in my use cases ( I have a lot of nested navigators )

```
import {
  NavigationActions
} from 'react-navigation'
import _ from 'lodash'

var previousAction = null;
var previousState = null;

export default (navigator) => {
  const navigateOnce = (getStateForAction) => (action, state) => {
    if (
      action && action.type === NavigationActions.NAVIGATE &&
      _.isEqual(previousAction, action) &&
      action.routeName === state.routes[state.index].routeName
    ) {
      return null
    } else {
      if (action && action.type === NavigationActions.NAVIGATE) {
        previousAction = action
        if (state) {
          previousState = state.routes[state.index]
        }
      }
    }

    return getStateForAction(action, state)
  }
  navigator.router.getStateForAction = navigateOnce(navigator.router.getStateForAction)
  return navigator
}
```

**Comment by maxdavidson** [comment]
This is how I solved it with TypeScript. It supports comparing routes with the same names but in different locations, by comparing their absolute paths.

```typescript
type Route<Param> = NavigationState | NavigationLeafRoute<Param> | NavigationStateRoute<Param>;

function hasNestedRoutes<Param>(route: Route<Param>): route is NavigationState | NavigationStateRoute<Param> {
  return (route as any).routes !== undefined;
}

function hasRouteName<Param>(route: Route<Param>): route is NavigationLeafRoute<Param> {
  return (route as any).routeName !== undefined;
}

function getCurrentRouteName<Param>(route: Route<Param>, prefix = ''): string {
  return hasNestedRoutes(route)
    ? getCurrentRouteName(route.routes[route.index], hasRouteName(route) ? `${route.routeName}/` : '')
    : `${prefix}${route.routeName}`;
}

function isNavigateAction(action?: AnyAction): action is NavigationNavigateAction {
  return action !== undefined && action.type === NavigationActions.NAVIGATE;
}

function getActionRouteName({ action, routeName }: NavigationNavigateAction, prefix = '') {
  return isNavigateAction(action) ? getActionRouteName(action, `${routeName}/`) : `${prefix}${routeName}`;
}

// Initial screen/route for the app
const initialState = router.getStateForAction(router.getActionForPathAndParams('Splash'), null);

export default function navigationReducer(
  state: NavigationState = initialState,
  action: NavigationAction,
): NavigationState {
  if (isNavigateAction(action) && getActionRouteName(action) === getCurrentRouteName(state)) {
    return state;
  }

  return router.getStateForAction(action, state) || state;
}
```

**Comment by javidjamae** [comment]
Thanks @gregblass! Your solution is very simple, avoids getting into the guts of react-navigation, and works like a charm. 

The react-navigation website says: `Easy-to-Use Navigators -- Start quickly with built-in navigators that deliver a seamless out-of-the box experience.` After spending 3 hours reading through this entire thread and evaluating different workarounds, I no longer believe that claim. 

Disallowing double tapping of a menu item should be the default behavior for any navigation library, no ifs, ands, or buts about it.

**Comment by pencilcheck** [comment]
I guess you can say they are "working towards" being easy to use and work out of box. But I agree at this current state it is not what it claims to be.

**Comment by andrewnaeve** [comment]
Even though there is no super clear parent child relationship when navigating to arbitrary screens, React Navigation provides a way to pass along props and methods as if these relationships did exist, by piggy-backing on the navigation props. Solutions with setTimeout or debouncing should be avoided. 

In this example, the RootComponent passes a method through this.props.navigation.state.params that will reset a boolean responsible for disabling your button. Because screens that aren't at the bottom of the stack navigator will be mounted and unmounted, you can use the second screen's componentWillUnmount to reset the value. 

```
class RootScreen extends Component {

    this.state = {
        disabled: false
    }

    resetDisabled = () => {
        this.setState({
            disabled: false
        })
    }

    handlePress = () => {
        this.setState({ disabled: true })
        this.props.navigation.navigate('SomeScreen', {
                resetDisabled: this.resetDisabled
	})
    }

    render() {
        return(
            <Button disabled={this.state.disabled}
        )
    }
}
```
```
class SomeScreen extends Component {

    componentWillUnmount() {
        const { navigation: { state: { params: { resetDisabled } } } } = this.props;
		resetDisabled();
        }
    ....
}
```

**Comment by koenpunt** [comment]
@amdrewnaeve state changes cause a render, which should be avoided while transitioning.

**Comment by koenpunt** [comment]
@rgabs FYI the module you created is already available as [`debounce`](https://www.npmjs.com/package/debounce), hence the term being used in this thread mutilple times.

**Comment by andrewnaeve** [comment]
@koenpunt I'm not sure that it does re-render while transitioning. Correct me, but navigating forward, the navigation is triggered by componentDidUpdate which triggers post render. On the way backward, componentWillUnmount isn't triggered until the the animation has taken the screen away from view,  _I think_.  In practice, my framerate looks the same with or without it, so I'm not sure it concerns me either way.

**Comment by jpca999** [comment]
Guys, have this issue ( Prevent navigating twice when clicking a button quickly )  been resolved in the latest version,  if yes, how to implement it without Redux ?

**Comment by patlux** [comment]
My solution for use without redux:

```javascript
import deepDiffer from 'react-native/lib/deepDiffer';
import { NavigationActions, StateUtils } from 'react-navigation';

export const getActiveRouteForState = navigationState =>
  navigationState.routes
    ? getActiveRouteForState(navigationState.routes[navigationState.index])
    : navigationState;

export const isEqualRoute = (route1, route2) => {
  if (route1.routeName !== route2.routeName) {
    return false;
  }

  return !deepDiffer(route1.params, route2.params);
};

const PATTERN_DRAWER_ROUTE_KEY = /^Drawer(Open|Close|Toggle)$/;
export const isDrawerRoute = route => PATTERN_DRAWER_ROUTE_KEY.test(route.routeName);

export const withNavigationPreventDuplicate = getStateForAction => {
  const defaultGetStateForAction = getStateForAction;

  const getStateForActionWithoutDuplicates = (action, state) => {
    if (action.type === NavigationActions.NAVIGATE) {
      const previousRoute = getActiveRouteForState(StateUtils.back(state));
      const currentRoute = getActiveRouteForState(state);
      const nextRoute = action;

      if (isDrawerRoute(currentRoute) && isEqualRoute(previousRoute, nextRoute)) {
        return StateUtils.back(state); // Close drawer
      }

      if (isEqualRoute(currentRoute, nextRoute)) {
        return null;
      }
    }

    return defaultGetStateForAction(action, state);
  };

  return getStateForActionWithoutDuplicates;
};

```

Then overwrite the `getStateForAction` method of your root navigator:

```javascript
const StackNavigation = StackNavigator({ ... });
StackNavigation.router.getStateForAction = withNavigationPreventDuplicate(
  StackNavigation.router.getStateForAction
);
```
:)

**Comment by swellingt0n** [comment]
@parakhod I tweaked your code a bit and it worked! Great fix for a redux implementation.

**Comment by acoulon99** [comment]
Glad I'm not the only one facing this issue :)
.. learning a lot looking through these work arounds

**Comment by paulpopiel** [comment]
I am doing some work on an existing project and noticed we had debounce wrapped around our `navigation.navigate('routeName', { params })` call.

It is however causing some slight lag on the navigation for us. 
Upon removing the debounce the navigation is instant.

I have subsequently done some testing and cannot replicate the 'multiple' navigation issue described in this thread... so I just wanted to confirm whether using `navigate` in this manner (i.e. just using routeName and params, with no action or redux use) causes this issue. 

Is it safe to remove the debounce in my case?

Thanks!

**Comment by soomin** [comment]
Solutions comparing route names don't work for back button.

I solved this by adding "from" property to navigation actions and by ignoring the same "from" navigation actions, which means they have been triggered in the same component instance.

This should work because every navigated screen has its unique own and child component instances and it cannot navigate to itself.

First, override the navigate and back action creators.
```javascript
const { navigate, back } = NavigationActions;
NavigationActions.navigate = ({ from, ...options }) => ({ ...navigate(options), from });
NavigationActions.back = ({ from, ...options }) => ({ ...back(options), from });
```

And, ignore the same "from" navigation actions in a reducer.
```javascript
const nextState = AppNavigator.router.getStateForAction(action, state);

if (nextState) {
  const { from } = action;
  nextState.from = from;

  if (from && from === state.from) {
    // navigation from the same component instance
    return state;
  }
}
```

Then, in any component, dispatch navigation actions like this..
```javascript
// Navigation
dispatch(NavigationActions.navigate({ routeName: 'ScreenName', from: this }));

// Back
dispatch(NavigationActions.back({ from: this }));
```

I hope this is helpful.

**Comment by phumaster** [comment]
I get code from @nihgwu #2 and edit to:

```javascript
if (action.type.startsWith('Navigation/')) {
    const {routeName} = action;
    const lastRoute = state.routes[state.routes.length - 1];
    if (routeName == lastRoute.routeName) return state
  }
  const nextState = StackBaseApp.router.getStateForAction(action, state);
  return nextState || state
```

It's works! :)

**Comment by Unforgiven-wanda** [comment]
I shall name my firstborn after you @patlux

**Comment by MacKentoch** [comment]
Why is it closed?

This behavior is clearly a 🐛 for a navigation library.

**Comment by Dror-Bar** [comment]
The solution @Doko-Demo-Doa suggested (where he edits the lib) doesn't seem to work for me.

Does anyone know of an updated solution to edit the lib and just add a delay / debounce for like 500 miliseconds every time I use navigate?

Also, why was this closed if it's still an issue? or is this considered a clicking issue and not a react-navigation issue?

**Comment by wellyshen** [comment]
Any progress for this issue officially?

**Comment by ivosabev** [comment]
+1 for an official solution

**Comment by ryanab** [comment]
I am having the same issue, but was playing around on a production applications from very large unnamed company that I believe uses React Native (Think Multiple Billion in Market Cap), and was able to reproduce the issue on there (and from my understanding they are using a different nav library); so I do not think the issue is specific to solely react-navigation, and may be more broad common than we think across various libs.

However that being said I would love an official solution, but in the interim I will post what I end up deciding on for anyone's future reference, but I think most of the redux solutions above look pretty good (I am not a fan of the timers personally but that is just personal opinion I am sure it works great).  

I am looking to see if I can do anything with the onTransitionStart  onTransitionEnd functions for those of us who are not currently routing all our navigation actions through redux, but ultimately this looks like the redux solution may be best option.

**Comment by AugustoAleGon** [comment]
@phumaster This solution works!

**Comment by LeonidVeremchuk** [comment]
OMG. So much workaround code ...
It is necessary to add something like - launchMode = "singleInstance"

**Comment by ericvicenti** [comment]
Hey folks, we just added a new feature: If you provide a key to the navigate action, you will always navigate to the same screen. So, if you use the following new API:

`this.props.navigation.navigate({ key: 'MyScreen1', routeName: 'ProfileScreen', params: {...}})`

You can call this action as many times as you like, and only one screen will be pushed! This is because the second time the action fires, it basically means "go to the existing route with key equal to 'MyScreen1'"

**Comment by Palisand** [comment]
Is this what we've all been waiting for?!

**Comment by ryanab** [comment]
@ericvicenti you are the man.  What NPM version is it?

**Comment by ericvicenti** [comment]
Its not on NPM yet, but we will get it released soon! If you want to test it out, you can point your package.json to github:

```
"react-navigation": "git://github.com/react-navigation/react-navigation.git#master"
```

**Comment by ryanab** [comment]
@ericvicenti confirmed to be working just tested beta 27 vs git master and it works!

Just a note to others reading (and please do not confuse this as a complaint I am very pleased this has just made my weekend much better 👍 ) but you can still fire off the navigation action twice if you press very fast.  However the good news is with this fix only one is actually on the stack.  If you are bothered by the double animation (and you have to press extremely fast its pretty tough) you may need to use a debounce etc. as discussed by some people above.  

Thanks again to everyone on react navigation team for this.

**Comment by ericvicenti** [comment]
Glad it works for you now!

If somebody wants to fix the double-tap behavior, maybe we could look into blocking the whole screen's touch events during a transition. Right now I think we block events to the screen, but the header still accepts touches

**Comment by nihgwu** [comment]
1 year passed 😅

**Comment by sun2rise** [comment]
yes @nihgwu 1 year passed, meanwhile  weeks ago i lost my job because of the bugs in the app and the slowness in the fixing... the navigation was quite complex and guess what library i was using?
This library is **TOXIC**, heavy-pollution for the react environment..

**Comment by ericvicenti** [comment]
I'm really sorry to hear that @sun2rise.. open source isn't easy 😞  Maybe you can help us de-toxify the library? What were the biggest issues for your app, aside from this one?

**Comment by ryanab** [comment]
I think it is a little unfair to blame a library which is open-source/free.  There are issues but it is a constantly changing and young project working with React Native which itself is still a very young and constabtly changing project, and should save you considerable time from rolling your own navigatir. That being said I have not done much to contribute and probably should; Eric if i were to look at the double tap issue where would be a good starting point / jumping off point.

**Comment by ericvicenti** [comment]
@ryanab, thanks for offering! In this case, I'd love to see an exploration of the core problem, and post a fresh issue with your findings. Here are the questions I have:

- Are these double-taps happening on screens, or in the header?
- There is something in StackNavigator that is supposed to block touches on the last screen once transition starts (PointerEventsContainer). Is this broken, or does it not happen fast enough?
- If the double-tap is in the header, can we use a similar technique to block pointer events during the transition?
- Maybe the button should be configured to only fire one event, even if it gets a double tap? Explore this approach, and discuss the drawbacks. Are there any cases where two quick taps really should result in two navigations? Maybe a "next" button?
- Maybe its ok for there to be two events, if we ignore the second one. This is the motivation for the new `key` field on `navigate`, which allows for an idempotent navigation. Maybe this could be improved to address the problem.

**Comment by Dror-Bar** [comment]
ericvicenti -
Thank you very much for this new feature, I think it is one of the most requested features.
Double-tapping quickly is sort of a general issue, which is amplified when dealing with navigations. Having a solution for this is **highly** appreciated.
That being said, is there any estimation for the release of beta 28?

**Comment by MacKentoch** [comment]
@fnicastri I clearly understand you frustration. I have multiple times little troubles that get me nerve. As you said there more than a coding issue right now, it is more about communication and visibility.

**But please**, keep in mind that people behind this take of their own time to try to make it better.

Wether we like or not this navigator we have to admit how hard work it is to make it working as anyone would like.

And don't forget how fast Facebook dropped old school but working Navigator in favor of community solution that were not ready for production.

I don't have time contribute to this navigator but I can at least say thank you everyone maintaining and making it grow every day.

**Comment by MacKentoch** [comment]
@fnicastri definitely agree

**Comment by ericvicenti** [comment]
@fnicastri, @MacKentoch,  thanks for the feedback! We're getting back up to speed, and tying to improve our practices here.

There are a ton of issues that get reported on this repo- closing the ones that don't make sense is our only way of keeping our sanity and focusing on the important problems.

> ignore serious bugs

What serious bugs do you see that are going ignored? We are trying to prioritize the most important and most serious bugs. Please help us by bringing them to our attention- especially when the issue is clearly written and something badly broken.

With regard to versioning, sorry about not documenting the changes very well. One problem is that we are stuck in this "beta" state where we cannot even respect semver, so we will hopefully release 1.0 soon, even if its not perfect, so we can start documenting major/minor/patch releases properly.

Again, your feedback is appreciated, and we're trying to improve the way things like this are handled, so keep on telling Brent and me how to improve!

**Comment by ryanab** [comment]
@ericvicenti will take a look at this most likely this weekend and open a new issue, but for now:

_Are these double-taps happening on screens, or in the header?
There is something in StackNavigator that is supposed to block touches on the last screen once transition starts (PointerEventsContainer). Is this broken, or does it not happen fast enough?
If the double-tap is in the header, can we use a similar technique to block pointer events during the transition?_

In the screen actually - will do some testing and try to find out if event is firing off before transition starts or if perhaps there is a bug (maybe something with nested navigators?).

_Maybe the button should be configured to only fire one event, even if it gets a double tap? Explore this approach, and discuss the drawbacks.
Are there any cases where two quick taps really should result in two navigations? Maybe a "next" button? How do we handle those cases, if at all?_

_Maybe its ok for there to be two events, if we ignore the second one. This is the motivation for the new key field on navigate, which allows for an idempotent navigation. Maybe this could be improved to address the problem._

If we are explicitly declaring our routes in a screen (i.e. not a next button in a stack), I do not really see any reason why we would want to ever allow a double tap, but perhaps there needs to be some type of setting in navigationOptions just in case?

Or, perhaps the answer is in the new key field navigate as you say.  Maybe if you are explicitly navigating using a key the second matching one should not only not get added to the stack as it does now with the new update, but we prevent the second transition from starting when the event is fired (again from your first question it seems like this is the intended default behavior).

Anyway, will open a new issue at some point soon but wanted to leave my thoughts here in the interim  in case anyone else is having similar issues they can track as well.

**Comment by gregblass** [comment]
Wait so 1.0 was released and this is still an open issue? Kind of blows my mind. Why is this marked as closed?

I just cannot understand why this would not be a very top of the list priority.

**Comment by pselibas** [comment]
Hi,
Was this issue not fixed in version 1.0.x? I was pretty sure I saw this on the roadmap, but I guess I was mistaken. 

While I have the utmost respect and appreciation for all the hard work and progress that has been made on this project, this is the one bug that I would consider a "show stopper" when making or promoting a quality react native app. 

Are there any non redux workarounds not involving a timer of some sort?
Keep up the excellent work and know that we are cheering you guys on.

**Comment by brentvatne** [comment]
https://github.com/react-navigation/rfcs/issues/16 for discussion, let's turn somehting into a rfc
